
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="This contains all material for the Introduction to NGS workshop.">
      
      
        <meta name="author" content="Sonika Tyagi">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.0, mkdocs-material-8.5.6">
    
    
      
        <title>Data Introduction - Introduction to ML for Biologists</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.20d9efc8.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../css/details.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#key-learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Introduction to ML for Biologists" class="md-header__button md-logo" aria-label="Introduction to ML for Biologists" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Introduction to ML for Biologists
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Data Introduction
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Introduction to ML for Biologists" class="md-nav__button md-logo" aria-label="Introduction to ML for Biologists" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Introduction to ML for Biologists
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../trainers/trainers.md" class="md-nav__link">
        The Trainers
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_3">
          Hands-on Modules
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Hands-on Modules" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Hands-on Modules
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Data Introduction
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Data Introduction
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#key-learning-outcomes" class="md-nav__link">
    Key Learning Outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resources-youll-be-using" class="md-nav__link">
    Resources You’ll be Using
  </a>
  
    <nav class="md-nav" aria-label="Resources You’ll be Using">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tools-used" class="md-nav__link">
    Tools Used
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#useful-links" class="md-nav__link">
    Useful Links
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#author-information" class="md-nav__link">
    Author Information
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fasta-format" class="md-nav__link">
    FASTA format
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fastq-format" class="md-nav__link">
    FASTQ format
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reading-in-the-sequence-file-in-fasta-format" class="md-nav__link">
    Reading in the sequence file in FASTA format
  </a>
  
    <nav class="md-nav" aria-label="Reading in the sequence file in FASTA format">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-using-a-custom-python-script" class="md-nav__link">
    1) Using a custom python script
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-using-biopython-utilities" class="md-nav__link">
    2) Using Biopython utilities
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../01_tokenization/handout/handout/" class="md-nav__link">
        Tokenization
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../02_RF_model/handout/handout/" class="md-nav__link">
        ML model
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../license.md" class="md-nav__link">
        License
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#key-learning-outcomes" class="md-nav__link">
    Key Learning Outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resources-youll-be-using" class="md-nav__link">
    Resources You’ll be Using
  </a>
  
    <nav class="md-nav" aria-label="Resources You’ll be Using">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tools-used" class="md-nav__link">
    Tools Used
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#useful-links" class="md-nav__link">
    Useful Links
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#author-information" class="md-nav__link">
    Author Information
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fasta-format" class="md-nav__link">
    FASTA format
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#fastq-format" class="md-nav__link">
    FASTQ format
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reading-in-the-sequence-file-in-fasta-format" class="md-nav__link">
    Reading in the sequence file in FASTA format
  </a>
  
    <nav class="md-nav" aria-label="Reading in the sequence file in FASTA format">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-using-a-custom-python-script" class="md-nav__link">
    1) Using a custom python script
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-using-biopython-utilities" class="md-nav__link">
    2) Using Biopython utilities
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>Data Introduction</h1>

<h2 id="key-learning-outcomes">Key Learning Outcomes</h2>
<p>After completing this practical the trainee should be able to:</p>
<ul>
<li>
<p>Recognize various genomics formats to represent DNA/RNA sequence data </p>
</li>
<li>
<p>Read in sequence data in FASTA format using Python or BioPython modules 
    decide on filters and cutoffs for cleaning up data ready for
    downstream analysis</p>
</li>
</ul>
<hr />
<h2 id="resources-youll-be-using">Resources You’ll be Using</h2>
<h3 id="tools-used">Tools Used</h3>
<p>BioPython SeqIO package:<br />
<a href="https://biopython.org/wiki/SeqIO">https://biopython.org/wiki/SeqIO</a>
<a href="https://biopython.org/docs/1.75/api/Bio.SeqIO.QualityIO.html">https://biopython.org/docs/1.75/api/Bio.SeqIO.QualityIO.html</a></p>
<p>Google Colaboratory:<br />
<a href="https://colab.research.google.com/">https://colab.research.google.com/</a></p>
<hr />
<h2 id="useful-links">Useful Links</h2>
<p>FASTA format: 
<a href="https://www.ncbi.nlm.nih.gov/genbank/fastaformat">https://www.ncbi.nlm.nih.gov/genbank/fastaformat</a></p>
<p>FASTQ format: 
<a href="https://www.ncbi.nlm.nih.gov/sra/docs/submitformats/#fastq-files">https://www.ncbi.nlm.nih.gov/sra/docs/submitformats/#fastq-files</a></p>
<hr />
<h2 id="author-information">Author Information</h2>
<p><em>Author(s):</em>  <br />
Sonika Tyagi, Sarthak Chauhan, Navya Tyagi, Tyrone Chen</p>
<hr />
<p>DNA sequencing of a biological specimen is performed to find the order of letters (or bases) A, C, G, and T in it to determine the code of DNA. Different organisms have DNA of different length, for example, human DNA is made of 3 billion pairs of individual bases. A highthroughput DNA sequencing experiment can generate miliions of copies of DNA sequences and these are provided as long text files. There are standard formats for presenting DNA sequence data. Two of the most common primary DNA data format are the FASTA (stands for FAST alignment) and FASTQ format.</p>
<h2 id="fasta-format">FASTA format</h2>
<p>A FASTA format of biological sequence has mainly two fields: a unique sequence identifier line that starts with a <code>&gt;</code> sign and the actuall sequence begins from the seconf line.</p>
<div class="highlight"><pre><span></span><code>&gt;My_seq_ID
TAATGGCTCT
GGAAGCTCT
TGGCTCTAGA
</code></pre></div>
<h2 id="fastq-format">FASTQ format</h2>
<p>Raw data from a highthroughput DNA sequencer is commonly stored in a FASTQ format. This format is similar to FASTA except it contains 4 different fiels and additionally have DNA quality information in it. Here, the sequence identifier line starts with an <code>@</code> sign, followed by the actual sequence, a line starting <code>+</code> can optionally ahve metadata informaiton, and finally a quality string of the ssame length as the DNA sequence.</p>
<div class="highlight"><pre><span></span><code>@SRR001666.1 071112_SLXA-EAS1_s_7:5:1:817:345 length=36
GGGTGATGGCCGCTGCCGATGGCGTCAAATCCCACC
+SRR001666.1 071112_SLXA-EAS1_s_7:5:1:817:345 length=36
IIIIIIIIIIIIIIIIIIIIIIIIIIIIII9IG9IC
</code></pre></div>
<p>The fouurth line int eh above format is a quality string where each letter represent a PHRED scale DNA base quality score reperesented on a scale of <code>0-40+</code>. 
The table below provides an indication to how quality scores related to base call accuracies.</p>
<table>
<thead>
<tr>
<th align="left">Phred Quality Score</th>
<th align="left">Probability of incorrect base call</th>
<th align="left">Base call accuracy</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">10</td>
<td align="left">1 in 10</td>
<td align="left">90%</td>
</tr>
<tr>
<td align="left">20</td>
<td align="left">1 in 100</td>
<td align="left">99%</td>
</tr>
<tr>
<td align="left">30</td>
<td align="left">1 in 1000</td>
<td align="left">99.9%</td>
</tr>
<tr>
<td align="left">40</td>
<td align="left">1 in 10,000</td>
<td align="left">99.99%</td>
</tr>
<tr>
<td align="left">50</td>
<td align="left">1 in 100,000</td>
<td align="left">99.999%</td>
</tr>
<tr>
<td align="left">60</td>
<td align="left">1 in 1,000,000</td>
<td align="left">99.9999%</td>
</tr>
</tbody>
</table>
<p><strong>Table 1:</strong> Phred quality scores are logarithmically linked to error probabilities</p>
<hr />
<h2 id="reading-in-the-sequence-file-in-fasta-format">Reading in the sequence file in FASTA format</h2>
<p>A file containing one or more DNA or RNA squences in FASTA format can be read using one of the following two methods. </p>
<h3 id="1-using-a-custom-python-script">1) Using a custom python script</h3>
<p>Open your Python notebook and type the following code in teh code cell and run.</p>
<div class="codehilite"><pre><span></span><code><span class="s s-Atom">def</span> <span class="nf">parse_fasta</span><span class="p">(</span><span class="s s-Atom">fh</span><span class="p">)</span><span class="o">:</span>
   <span class="s s-Atom">fs</span><span class="o">=</span><span class="p">[]</span>
   <span class="s s-Atom">fss</span><span class="o">=</span><span class="p">[]</span>
   <span class="s s-Atom">seq=&#39;&#39;</span>
   <span class="s s-Atom">for</span> <span class="s s-Atom">ln</span> <span class="s s-Atom">in</span> <span class="s s-Atom">fh</span><span class="p">:</span>
       <span class="s s-Atom">if</span> <span class="s s-Atom">ln</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="s s-Atom">==&#39;A&#39;</span> <span class="s s-Atom">or</span> <span class="s s-Atom">ln</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="s s-Atom">==&#39;T&#39;</span> <span class="s s-Atom">or</span> <span class="s s-Atom">ln</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="s s-Atom">==&#39;G&#39;</span> <span class="s s-Atom">or</span> <span class="s s-Atom">ln</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="s s-Atom">==&#39;C&#39;</span><span class="o">:</span>
          <span class="s s-Atom">seq</span><span class="o">=</span><span class="s s-Atom">seq</span><span class="o">+</span><span class="s s-Atom">ln</span><span class="p">[</span><span class="o">:-</span><span class="mi">2</span><span class="p">]</span>
       <span class="s s-Atom">else</span><span class="p">:</span>
          <span class="s s-Atom">fs</span><span class="p">.</span><span class="nf">append</span><span class="p">(</span><span class="s s-Atom">seq</span><span class="p">)</span>
          <span class="s s-Atom">seq=&#39;&#39;</span>
   <span class="s s-Atom">for</span> <span class="s s-Atom">element</span> <span class="s s-Atom">in</span> <span class="s s-Atom">fs</span><span class="p">:</span>
     <span class="s s-Atom">if</span> <span class="s s-Atom">element</span><span class="p">!</span><span class="s s-Atom">=&#39;&#39;</span><span class="o">:</span>
       <span class="s s-Atom">fss</span><span class="p">.</span><span class="nf">append</span><span class="p">(</span><span class="s s-Atom">element</span><span class="p">)</span>
     <span class="s s-Atom">else</span><span class="p">:</span>
       <span class="s s-Atom">fss</span><span class="o">=</span><span class="s s-Atom">fss</span>
   <span class="s s-Atom">return</span> <span class="s s-Atom">fss</span>

<span class="s s-Atom">fh</span><span class="o">=</span><span class="nf">open</span><span class="p">(</span><span class="s s-Atom">&#39;/Sequence.fasta&#39;</span><span class="p">)</span>
<span class="s s-Atom">input</span><span class="o">=</span><span class="nf">parse_fasta</span><span class="p">(</span><span class="s s-Atom">fh</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="s s-Atom">input</span><span class="p">)</span>
</code></pre></div>

<h3 id="2-using-biopython-utilities">2) Using Biopython utilities</h3>
<div class="highlight"><pre><span></span><code> from Bio import SeqIO
 for sequence in SeqIO.parse(&quot;example.fasta&quot;, &quot;fasta&quot;):
     print(sequence.id)
     print(sequence.seq)
     print (len(sequence)
</code></pre></div>
<p>A similar function is available to read the FASTQ format as well:</p>
<p><div class="highlight"><pre><span></span><code>from Bio import SeqIO
for record in SeqIO.parse(&quot;example.fastq&quot;, &quot;fastq&quot;):
    print(&quot;%s %s&quot; % (record.id, record.seq))
</code></pre></div>
Output will look like this:
<div class="highlight"><pre><span></span><code>SeqID1 CCCTTCTTGTCTTCAGCGTTTCTCC
SeqID2 TTGGCAGGCCAAGGCCGATGGATCA
SeqID3 GTTGCTTCTGGCGTGGGTGGGGGGG
</code></pre></div></p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../../.." class="md-footer__link md-footer__link--prev" aria-label="Previous: Home" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Home
            </div>
          </div>
        </a>
      
      
        
        <a href="../../../01_tokenization/handout/handout/" class="md-footer__link md-footer__link--next" aria-label="Next: Tokenization" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Tokenization
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; The authors; 2022-
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.5bf1dace.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.078830c0.min.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../../../js/details.js"></script>
      
    
  </body>
</html>