
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="This contains all material for the Introduction to NGS workshop.">
      
      
        <meta name="author" content="Sonika Tyagi">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.0, mkdocs-material-8.5.6">
    
    
      
        <title>Tokenization - Introduction to ML for Biologists</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.20d9efc8.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../css/details.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#preprocessing-biological-sequence-data-for-natural-language-processing" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Introduction to ML for Biologists" class="md-header__button md-logo" aria-label="Introduction to ML for Biologists" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Introduction to ML for Biologists
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Tokenization
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Introduction to ML for Biologists" class="md-nav__button md-logo" aria-label="Introduction to ML for Biologists" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Introduction to ML for Biologists
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../trainers/trainers.md" class="md-nav__link">
        The Trainers
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_3">
          Hands-on Modules
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Hands-on Modules" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Hands-on Modules
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../00_data_intro/handout/handout/" class="md-nav__link">
        Data Introduction
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Tokenization
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Tokenization
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning objectives:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resources-youll-be-using" class="md-nav__link">
    Resources You’ll be Using
  </a>
  
    <nav class="md-nav" aria-label="Resources You’ll be Using">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tools-used" class="md-nav__link">
    Tools Used
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#useful-links" class="md-nav__link">
    Useful Links
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#author-information" class="md-nav__link">
    Author Information
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    Workflow
  </a>
  
    <nav class="md-nav" aria-label="Workflow">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-split-each-input-sequence-into-tokens" class="md-nav__link">
    1) Split each input sequence into "tokens".
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-denoise-the-data-if-needed" class="md-nav__link">
    2) Denoise the data if needed
  </a>
  
    <nav class="md-nav" aria-label="2) Denoise the data if needed">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-numeric-encoding" class="md-nav__link">
    Step 1: Numeric encoding
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-transformation" class="md-nav__link">
    Step 2: Transformation
  </a>
  
    <nav class="md-nav" aria-label="Step 2: Transformation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#frequency-tables" class="md-nav__link">
    Frequency tables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#embedding-vector" class="md-nav__link">
    Embedding vector
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../../02_RF_model/handout/handout/" class="md-nav__link">
        ML model
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../license.md" class="md-nav__link">
        License
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-objectives" class="md-nav__link">
    Learning objectives:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#resources-youll-be-using" class="md-nav__link">
    Resources You’ll be Using
  </a>
  
    <nav class="md-nav" aria-label="Resources You’ll be Using">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#tools-used" class="md-nav__link">
    Tools Used
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#useful-links" class="md-nav__link">
    Useful Links
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#author-information" class="md-nav__link">
    Author Information
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    Workflow
  </a>
  
    <nav class="md-nav" aria-label="Workflow">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-split-each-input-sequence-into-tokens" class="md-nav__link">
    1) Split each input sequence into "tokens".
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-denoise-the-data-if-needed" class="md-nav__link">
    2) Denoise the data if needed
  </a>
  
    <nav class="md-nav" aria-label="2) Denoise the data if needed">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-numeric-encoding" class="md-nav__link">
    Step 1: Numeric encoding
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-transformation" class="md-nav__link">
    Step 2: Transformation
  </a>
  
    <nav class="md-nav" aria-label="Step 2: Transformation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#frequency-tables" class="md-nav__link">
    Frequency tables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#embedding-vector" class="md-nav__link">
    Embedding vector
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="preprocessing-biological-sequence-data-for-natural-language-processing">Preprocessing Biological Sequence Data for Natural Language Processing</h1>
<h2 id="learning-objectives">Learning objectives:</h2>
<p>The overall objective is to understand how biological sequence data is transformed into a machine-readable format for input into a machine learning pipeline.</p>
<ul>
<li>View the flow of data through a conventional DNA data preproceessing pipeline</li>
<li>Familiarity with common genomic data transformations and representations</li>
<li>Explore how large sequences are separated into smaller components</li>
<li>Awareness of the ecosystem of available libraries</li>
</ul>
<hr />
<h2 id="resources-youll-be-using">Resources You’ll be Using</h2>
<h3 id="tools-used">Tools Used</h3>
<p>Python Package:<br />
Pandas, BioPython, sklearn, gensim</p>
<p>BioPython SeqIO package:<br />
<a href="https://biopython.org/wiki/SeqIO">https://biopython.org/wiki/SeqIO</a></p>
<p>Google Colaboratory:<br />
<a href="https://colab.research.google.com/">https://colab.research.google.com/</a></p>
<hr />
<h2 id="useful-links">Useful Links</h2>
<hr />
<h2 id="author-information">Author Information</h2>
<p><em>Primary Author(s):</em>  <br />
Tyrone Chen, Navya Tyagi, Naima Vahab</p>
<p><em>Contributor(s):</em>  <br />
Sonika Tyagi</p>
<hr />
<h2 id="workflow">Workflow</h2>
<p>Assuming you have obtained well-defined sequence data and their corresponding metadata:
1. Split each input sequence into &ldquo;k-mers&rdquo; or &ldquo;tokens&rdquo;.
2. Denoise the data if needed.
3. Convert these tokens into a numeric representation.
4. Transform the numeric representation as needed, depending on your use case.</p>
<blockquote>
<p><em>NOTE</em>: in the context of this tutorial, we will be using the term <code>token</code> and <code>k-mer</code> interchangeably.</p>
</blockquote>
<p>Let us take a single biological sequence as an example. Assume that this corresponds to a gene of interest:</p>
<div class="highlight"><pre><span></span><code>input_sequence = &quot;TAATGGCTCT&quot;
</code></pre></div>
<h3 id="1-split-each-input-sequence-into-tokens">1) Split each input sequence into &ldquo;tokens&rdquo;.</h3>
<p>It is impractical in most cases for the machine to obtain a representation of a whole sequence, for example a whole document or chromosome. On the other hand, the same is true for obtaining individual letters of text or genomic DNA. While in the case of some languages such as English, it is possible to exploit the property of space-separation, this is not applicable to genomic DNA. Therefore, in most cases an arbitrary-length split is performed.</p>
<div class="highlight"><pre><span></span><code>def get_kmers(sequence: str, length: int):
  &quot;&quot;&quot;
  Take a dna sequence as input and split the sequence into k-mers / tokens
  &quot;&quot;&quot;
  return [sequence[x:x+length].upper() for x in range(len(sequence) - length + 1)]

# in this example, we will split on length 3 k-mers
length = 3
kmers = get_kmers(input_sequence, length)
print(kmers)
</code></pre></div>
<p>We can see that we get k-mers or tokens of length 3 in our example. It is important to note that we take a sliding window of k-mers:</p>
<div class="highlight"><pre><span></span><code>INPUT : TAATG...
KMER_1: TAA--
KMER_2: -AAT-
KMER_3: --ATG
</code></pre></div>
<blockquote>
<p><em>NOTE</em>: in real applications, a range of longer k-mers are commonly used, and this is not the only way to split your data into chunks0</p>
</blockquote>
<h3 id="2-denoise-the-data-if-needed">2) Denoise the data if needed</h3>
<p>In some cases, we want to remove certain tokens, commonly tokens with low information content. In English, this is straightforward and involves filtering out &ldquo;stopwords&rdquo;, for example <code>the</code>, <code>and</code> or <code>is</code>. In Biology, this is not always done, and will require review on a case-by-case basis.</p>
<div class="highlight"><pre><span></span><code>def filter_kmers(tokens: str, stopwords: list):
  &quot;&quot;&quot;
  Take an input dna sequence and list of stopwords to remove from the data.
  &quot;&quot;&quot;
  return [x for x in tokens if x not in stopwords]

# in this example, let us pretend this list of k-mers have low information content
stopwords = [&quot;TAA&quot;, &quot;AAT&quot;]
filtered_kmers = filter_kmers(kmers, stopwords)
print(filtered_kmers)
</code></pre></div>
<h4 id="step-1-numeric-encoding">Step 1: Numeric encoding</h4>
<ul>
<li>Ordinal encoding</li>
<li>One-hot encoding</li>
</ul>
<p>The first type of text encodings are ordinal encodings. With this, individual <code>tokens</code> are assigned an integer value as shown below:</p>
<div class="highlight"><pre><span></span><code>from sklearn.preprocessing import OrdinalEncoder

def encode_ordinal(input_sequence: str, length: int):
  &quot;&quot;&quot;
  Take a list of k-mers and perform ordinal encoding
  &quot;&quot;&quot;
  tokens = [[x] for x in get_kmers(input_sequence, length)]
  encoder = OrdinalEncoder()
  return encoder.fit_transform(tokens)

ordinal = encode_ordinal(input_sequence, length)
print(ordinal)
</code></pre></div>
<div class="highlight"><pre><span></span><code>[[5.]
 [0.]
 [1.]
 [7.]
 [4.]
 [3.]
 [2.]
 [6.]]
</code></pre></div>
<blockquote>
<p><em>NOTE</em>: As an alternative to ordinal encoding, it is also possible to encode tokens as one-hot encodings. This method is not practical to use in most cases due to its sparsity, but we demonstrate an example for completeness. Compare this to the above:</p>
</blockquote>
<div class="highlight"><pre><span></span><code>from numpy import asarray
from sklearn.preprocessing import OneHotEncoder

def onehoten(sequence,size):
  r&quot;&quot;&quot;Array of categories(here string) are sorted and returns binary variables for each category&quot;&quot;&quot;
  doc=sentence_in_list(sequence,size)   
  data = asarray(doc)
  encoder = OneHotEncoder(sparse=False)
  onehot = encoder.fit_transform(data)
  return onehot
onehoten(input_sequence,length)
</code></pre></div>
<div class="highlight"><pre><span></span><code>[[ 0 0 0 0 0 1 0 0]
 [ 1 0 0 0 0 0 0 0]
 [ 0 1 0 0 0 0 0 0]
 [ 0 0 0 0 0 0 0 1]
 [ 0 0 0 0 1 0 0 0]
 [ 0 0 0 1 0 0 0 0]
 [ 0 0 1 0 0 0 0 0]
 [ 0 0 0 0 0 0 1 0]]
</code></pre></div>
<p>While these can be used on their own, note that this is often just the first step, and that ordinal encodings can be transformed further.</p>
<p>While these can be used on their own, note that this is often just the first step, and that ordinal encodings can be transformed further.</p>
<h4 id="step-2-transformation">Step 2: Transformation</h4>
<ul>
<li>Frequency tables</li>
<li>Embedding vector</li>
<li>Positional encoding</li>
</ul>
<h5 id="frequency-tables">Frequency tables</h5>
<p><strong>Count vectorisation</strong></p>
<p>First, each token is counted and these values are assigned to a dictionary. On its own, these counts can be vectorised directly. However, it is often practical to weight tokens by their frequency as shown in the next step:</p>
<p><strong>TF-IDF (Term Frequency-Inverse Document Frequency)</strong></p>
<p>TF-IDF is analogous to a library-size correction in conventional short-read RNA-Seq, allowing different size sequence collections to be compared on similar scales. It is computed on the previous counts prior to vectorisation.</p>
<p>TF-IDF can be calculated as follows:</p>
<div class="highlight"><pre><span></span><code>tf-idf(t, d) = tf(t, d) * log(N/(df + 1))
</code></pre></div>
<p>Where:
- t = token
- d = sequence
- N = total number of sequences in the library</p>
<p>Within <code>scikit-learn</code>, the counting process is abstracted and not shown directly. We demonstrate this intermediate step for reference:</p>
<p><div class="highlight"><pre><span></span><code># input
&#39;TAA AAT ATG TGG GGC GCT CTC TCT&#39;

# intermediate (unweighted matrix)
{&#39;TAA&#39;: 1, &#39;AAT&#39;: 1, &#39;ATG&#39;: 1, &#39;TGG&#39;: 1, &#39;GGC&#39;: 1, &#39;GCT&#39;: 1, &#39;CTC&#39;: 1, &#39;TCT&#39;: 1}

# resulting matrix
</code></pre></div>
<div class="highlight"><pre><span></span><code>from sklearn.feature_extraction.text import TfidfVectorizer

def tfidf(input_sequence: str, length: int):
  &quot;&quot;&quot;
  Take a dna sequence and k-mer size as input, 
  output a matrix of TF-IDF features.
  &quot;&quot;&quot;
  data = [&quot; &quot;.join(get_kmers(input_sequence, length))]
  tfidf = TfidfVectorizer()
  tfidf = tfidf.fit_transform(data)
  return tfidf.toarray()

tfidf_matrix = tfidf(input_sequence,length) 
print(tfidf_matrix)
</code></pre></div></p>
<p>Although the frequency-based approaches above are often effective on their own, one key piece of information is lost. Individual tokens are treated as independent occurrences, which is not realistic in both natural language and biology. To account for the semantic relationship between tokens, a different approach is required.</p>
<h5 id="embedding-vector">Embedding vector</h5>
<p>A latent space is generated from all tokens in the corpus. This n-dimensional <code>embedding</code> accounts for every combination of tokens in the data. Taking a single sequence (i.e. group of tokens) and projecting this back onto the embedding returns a unique vector for each sequence. Examining these further shows that tokens which are closely related to each other are naturally recapitulated by the model. For example, in a real life English dataset, you may expect the following word pairs to be closely related to each other:</p>
<p><br>
<img alt="image" src="../../images/embed.png" /></p>
<div class="highlight"><pre><span></span><code>import gensim
from gensim.models import Word2Vec

def create_word2vec(input_sequence: str, length: int):
  &quot;&quot;&quot;
  Input a list of tokens to generate an embedding of word vectors
  &quot;&quot;&quot;
  data = get_kmers(input_sequence, length)
  return Word2Vec([data], min_count=1)

def map_token(input_sequence: str, model: gensim.models.base_any2vec):
  &quot;&quot;&quot;
  Input a token and embedding and map the token onto the embedding
  &quot;&quot;&quot;
  return model[input_sequence]

embedding = create_word2vec(input_sequence, length)
vector = map_token(&#39;TAA&#39;, embedding)
print(vector)
</code></pre></div>
<p>While we demonstrate <code>gensim</code> as one example, many embedding algorithms exist. A comprehensive overview of each method is out of the scope of this tutorial, but we list a few popular methods for your reference:</p>
<ul>
<li>Word2Vec</li>
<li>dna2vec (dna equivalent of the above)</li>
<li>GloVe</li>
</ul>
<h1 id="end">END</h1>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../../../00_data_intro/handout/handout/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Data Introduction" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Data Introduction
            </div>
          </div>
        </a>
      
      
        
        <a href="../../../02_RF_model/handout/handout/" class="md-footer__link md-footer__link--next" aria-label="Next: ML model" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              ML model
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; The authors; 2022-
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.5bf1dace.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.078830c0.min.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../../../js/details.js"></script>
      
    
  </body>
</html>